test/stdlib/OSLogCheckLogs.swift

```diff
@@ -0,0 +1,192 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift %s -swift-version 5 -o %t/OSLogCheckLogs
+// RUN: %target-run %t/OSLogCheckLogs | %FileCheck %s
+// REQUIRES: executable_test
+// REQUIRES: OS=macosx
+
+// Run-time tests that check that the messages logged with the existing os_log
+// APIs and the new prototype APIs (based on string interpolation) are correctly
+// readable from the log archive. These tests log some messages and execute the
+// `log show` command (by forking a process) to read back the logs and display
+// them.
+
+// If these tests fail, it means either there are errors in the overlay
+// implementation or certain guarantees that os_log relies on, such as that the
+// format string passed to the C ABIs point to a string literal in the __cstring
+// section of compiled binary, is broken by the compiler.
+
```

Comment:
You can also set `stdout` on Process, which should allow you to check this output programmatically rather than through FileCheck (which is a little looser in its checks).

Reply:
I see. One thing I like about this that the CHECKs are next to the `log` calls and we can use regular expressions and other functionality of FileCheck. 

However, I am trying to debug the test error and why the logs are not available in the remote testing machine. I suspect I may have to explicitly set stdout  of Process to redirect to a file probably.

Reply:
What I was envisioning was something like (total straw man, didn't even try to compile):

```swift
extension String {
  func lines() -> [String] {
    return self.split(separator: "\n").map { String($0) }
  }
}
struct LogChecks {
  private let subsystem: String
  private let startTime: Int64
  private var messages: [(String, file: StaticString, line: UInt)] = []

  private func validateLogs() {
    ... dumpLogs
    ... Check line-by-line with `expectEqual`, feeding in file/line
  }

  func log(
    _ logF: @autoclosure () -> (), expecting: String,
    _ file: StaticString = #file, _ line: UInt = #line
  ) {
    logF()
    let entries = expecting.lines().map { ($0, file, line) }
    messages.append(contentsOf: entries)
  }

  static func test(_ subsystem: String, _ f: (inout LogChecks) -> ()) {
    var checks = LogChecks(
      subsystem: subsystem, startTime: Int64(Date().timeIntervalSince1970))
    f(&checks)
    checks.validateLogs()
  }
}

OSLogTestSuite.test("log messages and check logs") {
  LogChecks.test("com.apple.oslog.test") { logCheck in
    logCheck.log(
      os_log("test: %{public}d", log: logObject, 42),
      expecting: "42")
     ...
  }
}

```

You can of course not do the closure thing and instead have `defer { logCheck.validateLogs() } ` or any formulation. The `lines()` is assuming that logging will smash the grouping of multi-line output, if not, then even better and you can drop it.

Reply:
An additional advantage is being able to run programmatic tests, since you can programmatically generate the `expecting` entry but you can't do that for CHECK lines. You can also use multi-line string literals and interpolation. E.g. the newer log invocations take interpolated literals, and the expecting line could be the same (or similar) interpolated literal, but using `String`'s interpolation.

Reply:
I see what you mean. Thanks for the straw man code. Just one comment: 

>  ... Check line-by-line with `expectEqual`, feeding in file/line

This is going to involve doing some inexact match e.g. like `logLine.contains(expectedString)`. Ideally, the expected string should be a regular expression, but all of that feature is not needed here, expect in one case where I had to check that there is nothing that succeeds the expected string.  Anyway, that is an implementation detail.

Reply:
No knowing anything about how the logging output looks, why do you want to do partial matches?

Reply:
It just some has some details that precede the message. E.g. 

Timestamp                       Thread     Type        Activity             PID    TTL  
 2019-05-22 15:00:57.186528-0700 0x57b25ba  Default     0x0                  60835  0    OSLogCheckLogs: [com.apple.oslog.test:LogCheck] test: 42

Reply:
Ah, so `logLine.hasSuffix(expectedLine)` then, which will also show you that nothing follows the expected string.

Reply:
Or may be we can strip the log lines of this meta data before matching